echopop/workflow/feat_hake.py

```diff
@@ -163,124 +168,482 @@
 )
 
+# ==================================================================================================
+# Fit length-weight regression to the binned data
+# -----------------------------------------------
+length_weight_dataset: pd.DataFrame = dict_df_bio_binned_ks["specimen"].copy()
 
-# ===========================================
-# Compute biological composition based on stratum
-length_bins: np.array  # length bin specification
-df_length_weight, df_regression = get_proportions.length_weight_regression(
-    dict_df_bio["specimen"], length_bins
+# Outputs: 
+# ---- `pandas.Series` that includes the fitted slope and intercepts for the length-weight 
+# ---- log-linear regressions. This can either be ran directly on a `pandas.DataFrame` or via 
+# ---- `pandas.DataFrame.groupby`
+length_weight_coefs_all: pd.Series = biology.fit_length_weight_regression(length_weight_dataset)
+length_weight_coefs_sex = length_weight_dataset.groupby(["sex"]).apply(
+    biology.fit_length_weight_regression,
+    include_groups=False
 )
-# df_regression seems unused afterwards -- good as a record?
 
-# Get counts ----------------
-df_aged_counts = get_proportions.fish_count(  # previously "aged_number_distribution"
-    df_specimen=dict_df_bio["specimen"],
-    df_length=dict_df_bio["length"],
-    aged=True,
-    sexed=True,
+# ==================================================================================================
+# Compute the mean weights per length bin
+# ---------------------------------------
+data: pd.DataFrame = dict_df_bio_binned_ks["specimen"].copy()
+regression_coefficients: Union[pd.Series, pd.DataFrame] = length_weight_coefs_all
+impute_bins: bool = True
+minimum_count_threshold: int = 5
+
+# Outputs: 
+# ---- `pandas.DataFrame` with the fitted mean weights per length bin with the corresponding 
+# ---- index column, if it exists (e.g. `"sex"`)
+
+# All fish
+binned_weights_df_all = biology.length_binned_weights(
+    data, length_distribution, regression_coefficients, impute_bins, minimum_count_threshold
 )
-df_unaged_counts = get_proportions.fish_count(  # previously "unaged_number_distribution"
-    df_specimen=dict_df_bio["specimen"],
-    df_length=dict_df_bio["length"],
-    aged=False,
-    sexed=True,
+
+# Sex-specific
+binned_weights_df_sexed = biology.length_binned_weights(
+    data=dict_df_bio_binned_ks["specimen"], 
+    length_distribution=length_distribution,
+    regression_coefficients=length_weight_coefs_sex, 
+    impute_bins=True, 
+    minimum_count_threshold=5
 )
-# Previously there was also "aged_number_distribution_filtered"
-# but it is simply df_aged_counts with unsexed fish removed,
-# I think it is better to have that explicitly in the code,
-# so removed OUTSIDE of the get_fish_count function
 
+# Concatenate the sex-specific and all-specimen dataframes
+binned_weight_table = pd.concat(
+    [binned_weights_df_sexed.copy(),
+    binned_weights_df_all.assign(sex="all")],
+    ignore_index=True
+)
```

Comment:
Why can these just be a single dataframe with multiple columns, each being the male, female, and all binned weights?

Reply:
I think it would be good to get this resolved once and for all, since this is refactoring and the resulting code should be modularized and as clean as possible.